### PR TITLE
costa: add the plugin not responding detection

### PR DESCRIPTION
**Pull request by yorkie**
This is able to tell the user if some plugin not responding on loading stage, and this is to implement the handler at https://github.com/alibaba/pipcook/issues/631#issuecomment-729560824.

### CODE REVIEW

**Comment by yorkie** [comment]
This PR is to send a new WRITE message "plugin loaded" to notify if the plugin is loaded before a plugin is completed.

Note that, to send the message before the "pong", this PR introduces a new WRITE message "emit" which is used to send the message before "pong", and it adds the handlers reset when `sendAndWait()` is done.

**Comment by yorkie** [comment]
The test case is WIP.

File: packages/costa/src/client/entry.ts [hunk 1]
```diff
@@ -24,6 +24,18 @@ function recv(respOp: PluginOperator, ...params: string[]): void {
   }));
 }
 
+/**
+ * Send a `emit` message back from the client process, this will not stop the message handler.
+ * @param respOp the operator of response.
+ * @param msg
+ */
+function emit(respOp: PluginOperator, msg: string): void {
+  process.send(PluginProtocol.stringify(respOp, {
```
Comment (FeelyChau) [inline packages/costa/src/client/entry.ts:33]: Should catch the error from callback?

File: packages/costa/src/protocol.ts [hunk 2]
```diff
@@ -20,7 +20,7 @@ export interface PluginResponse {
 export enum PluginOperator {
   START = 0x30,
   WRITE = 0x31,
-  READ = 0x32
+  READ = 0x32,
```
Comment (FeelyChau) [inline packages/costa/src/protocol.ts:23]:
```suggestion
  READ = 0x32
```

File: packages/costa/src/runnable.test.ts [hunk 3]
```diff
@@ -119,4 +121,18 @@ describe('start runnable in normal way', () => {
...
+
+  it('should destroy the runnable when it blocks on loading plugin', async () => {
+    const logger = process;
+    const costa = new CostaRuntime(opts);
+    const r2 = new PluginRunnable(costa, logger);
+    await r2.bootstrap({ logger, pluginNotRespondingTimeout: 500 });
+    const simple = await costa.fetch(path.join(__dirname, '../tests/plugins/nodejs-not-responding'));
+    await costa.install(simple, process);
+    try {
+      await r2.start(simple, { foobar: true });
```
Comment (FeelyChau) [inline packages/costa/src/runnable.test.ts:133]: We could use `expectAsync` instead to make sure the asynchronous error has been caught.

File: packages/costa/src/runnable.ts [hunk 4]
```diff
@@ -155,12 +169,36 @@ export class PluginRunnable {
...
       compPath + `/node_modules/${pkg.name}`);
 
-    const resp = await this.sendAndWait(PluginOperator.WRITE, {
-      event: 'start',
-      params: [
-        pkg,
-        ...args
-      ]
+    // wrap the PNR with the start logic.
+    const resp = await new Promise<PluginMessage>(async (resolve, reject) => {
```
Comment (FeelyChau) [inline packages/costa/src/runnable.ts:173]: I think using `await` in `Promise` is not a good practice.

**Review by FeelyChau — CHANGES_REQUESTED** [review]

File: packages/costa/src/runnable.ts (hunk 4 above)
Comment (yorkie) [reply packages/costa/src/runnable.ts:173]: Is there a better way this?

File: packages/costa/src/client/entry.ts (hunk 1 above)
Comment (yorkie) [reply packages/costa/src/client/entry.ts:33]: There might not be a callback here :(

File: packages/costa/src/runnable.ts [hunk 5]
```diff
@@ -244,22 +276,42 @@ export class PluginRunnable {
...
    * @param msg
    */
-  private async sendAndWait(op: PluginOperator, msg: PluginMessage): Promise<PluginMessage> {
-    await this.send(op, msg);
-    debug(`sent ${msg.event} for ${this.id}, and wait for response`);
+  private async sendAndWait(op: PluginOperator, msg: PluginMessage, timeout?: number): Promise<PluginMessage> {
+    return new Promise<PluginMessage>((resolve, reject) => {
+      let opNotRespondingTimer: NodeJS.Timer;
+      if (typeof timeout === 'number' && timeout > 0) {
+        opNotRespondingTimer = setTimeout(() => {
```
Comment (FeelyChau) [inline packages/costa/src/runnable.ts:283]: I think we could insert the timer into function `send` like this:
```js
  /**
   * Reads the message, it's blocking the async context util.
   * @params timeout timeout value in millisecond
   */
  private async read(timeout = 0): Promise<PluginProtocol> {
    return new Promise((resolve, reject) => {
     let timer;
      if (timeout > 0) {
        timer = setTimeout(() => {
          reject(new Error('read timeout'));
        }, timeout);
      }
      this.onread = (res: PluginProtocol) => {
        if (timer) {
          clearTimeout(timer);
        }
        resolve(res);
      };
      this.onreadfail = (err: Error) => {
        if (timer) {
          clearTimeout(timer);
        }
        reject(err);
      };
    });
  }
```
Is this a little more concise?

**Review by FeelyChau — CHANGES_REQUESTED** [review]

**Comment by yorkie** [comment]
@FeelyChau The CI is green now :)

**Review by FeelyChau — APPROVED** [review]

**Review by WenheLI — APPROVED** [review]
LGTM

**Pull request merged**